scanapi/evaluators/code_evaluator.py

```diff
@@ -55,12 +54,97 @@ def evaluate(cls, sequence, spec_vars, is_a_test_case=False):
         except Exception as e:
             raise InvalidPythonCodeError(str(e), code)
 
+    @classmethod
+    def _get_allowed_modules(cls):
+        """Dynamically import allowed modules.
+
+        Returns:
+            dict: Dictionary of module names to imported modules
+        """
+        return {name: __import__(name) for name in cls.ALLOWED_MODULES}
+
+    @classmethod
+    def _get_safe_globals(cls, response=None):
+        """Create a secure global context for code execution.
+
+        Args:
+            response: Optional response object for test assertions
+
+        Returns:
+            dict: Safe global context with restricted access
+        """
+        safe_context = safe_globals.copy()
+        safe_context["__builtins__"] = safe_builtins.copy()
+
+        # Add iterator functions for generator expressions and comprehensions
+        safe_context["_iter_unpack_sequence_"] = iter
+        safe_context["_getiter_"] = iter
+        safe_context["_getattr_"] = getattr
+        # Enables obj[key] access
+        safe_context["_getitem_"] = lambda obj, key: obj[key]
+        essential_builtins = {
+            "all": all,
+            "any": any,
+            "len": len,
+            "str": str,
+        }
+        safe_context["__builtins__"].update(essential_builtins)
+
+        # Add allowed modules via dynamic import
+        allowed_modules = cls._get_allowed_modules()
+        safe_context.update(allowed_modules)
+
+        # Add response object if provided (for test assertions)
+        if response is not None:
+            safe_context["response"] = response
+
+        return safe_context
+
+    @classmethod
+    def _safe_eval(cls, code, global_context=None):
+        """Safely evaluate Python code using RestrictedPython with mode='eval'.
+
+        Args:
+            code[string]: Python code to evaluate
+            global_context[dict]: Global context for evaluation
+
+        Returns:
+            Result of code evaluation
+
+        Raises:
+            InvalidPythonCodeError: If code compilation or execution fails
+        """
+        if global_context is None:
+            global_context = cls._get_safe_globals()
+
+        try:
+            # Strip whitespace to avoid IndentationError in RestrictedPython
+            clean_code = code.strip()
+
+            # Compile the code with restrictions using mode='eval'
+            compiled_code = compile_restricted(
+                clean_code, "<string>", mode="eval"
+            )
+            if compiled_code is None:
+                raise InvalidPythonCodeError(
+                    "Failed to compile restricted code", code
+                )
+
+            # Execute the compiled code securely
```

Comment:
I know this wasn’t part of the previous implementation, but it might be a good improvement to log the exact command that will run. That way, it’s easier to debug and understand what’s being executed without having to dig into the code.